szar/clustering.py

```diff
@@ -130,78 +130,105 @@ def fine_sfunc(self, fsky, nsubsamples):
         zs = self.HMF.zarr
         zgridedges = self.HMF.zarr_edges
 
-        values = np.empty(zs.size)
+        fine_zgrid = np.empty((zs.size, nsubsamples))
         for i in range(zs.size):
-            if i == 0:
-                fine_zs = np.linspace(zs[i], zgridedges[i+1], nsubsamples)
-            elif i == zs.size - 1:
-                fine_zs = np.linspace(zgridedges[i], zs[i])
-            else:
-                fine_zs = np.linspace(zgridedges[i], zgridedges[i+1], nsubsamples)
-            ntils = self.ntilde_interpol(fine_zs)
-            dvdz = self.dVdz_fine(fine_zs)
-            integral = simps(dvdz * ntils**2, fine_zs)
-            values[i] = integral * 4 * np.pi * fsky
-        return values
+            fine_zgrid[i,:] = np.linspace(zgridedges[i], zgridedges[i+1], nsubsamples)
+
+        fine_zgrid = fine_zgrid[1:-1]
+
+        ntils = self.ntilde_interpol(fine_zgrid)
+        dvdz = np.array([self.dVdz_fine(zs) for zs in fine_zgrid])
+
+        dz = fine_zgrid[0,1] - fine_zgrid[0,0]
+
+        assert np.allclose(dz * np.ones(tuple(np.subtract(fine_zgrid.shape, (0,1)))),  np.diff(fine_zgrid,axis=1), rtol=1e-3)
+
+        integral = simps(dvdz * ntils**2, dx=dz)
+        integral *= 4 * np.pi * fsky
+
+        return integral
```

Comment:
Did you test this new volume integral?

Reply:
Yes, I re-ran the tests for V_eff and V_0 and the changes pass (and in fact make our plots of V_eff closer to that of the paper we were comparing to - see Nuclino for this).